docs/proposals/resource-quota.md

```diff
@@ -0,0 +1,453 @@
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Resource Quota
+
+## Background
+
+This document describes proposed enhancements to `ResourceQuota` along with a
+phased set of development milestones.
+
+## Use Cases
+
+* As a cluster administrator, I want the ability to quota:
+ * resources per namespace to prevent resource hogs
+ * resources per kind to control cluster size
+ * minimum compute resources allowed
+ * maximum compute resources allowed
+ * compute resources for terminating vs non-terminating workloads to fairly share nodes
+ * compute resources for best-effort vs non-besteffort pods to fairly share nodes
+ * creation rate per kind to prevent abuse 
+* As a namespace administrator, I want the ability to quota:
+ * resource consumption based on local understanding of individual workload
+
+## High-level description
+
+This proposal augments the existing API to modify a `ResourceQuota`.
+
+### Ability to quota the min vs the max compute resources
+
+The existing `ResourceQuota` document constrains the total amount of compute
+resource requests.  This is useful when cluster-managers are interested in
+controlling explicit resource guarantees often associated with higher SLA.
+The end-user creating the pod is expected to have intimate knowledge on 
+their minimum required resource as well as their potential limits.
+
+There are many environments where cluster-managers do not extend this level
+of trust to their end-user because user's often request too much resource, and
+they have trouble reasoning between what they hope to have available for their
+application versus what their application actually needs.  In these environments,
+cluster-managers often just expose a single value (the limit) to the end-user,
+and internally, they set the request as a factor of the limit to control for 
+over-commit or based on known-historical trends.  In this environment, the
+cluster-manager is interested in exposing a quota to their end-users that maps
+to their desired limit instead of their request.
+
+To facilitate both operational environments, we propose a `ResourceQuota` can
+have a `ResourceAccountingPolicy`.  If set to `Requests`, the quota document will charge
+for quota usage based on the `resources.requests` on a `Kind`.  If set to `Limits`,
+the quota document will charge for quota usage based on `resources.limits` on a `Kind`.
+If not specified, it will default to `Requests` to preserve backward-compatibility.
+If set to `Limits`, a `Kind` must specify an upper limit for each resource tracked
+by quota.  The `ResourceAccountingPolicy` is only pertinent to resources that support
+bursting semantics.
+
+### Filter the set of resources tracked
+
+The set of resources tracked by a `ResourceQuota` can be filtered
+via an optional `LabelSelector` and `FieldSelector`.
```

Comment:
I just realized that updating multiple matching ResourceQuota objects might require multi-object atomic transactions, which we don't have.  Can you confirm or deny, Derek? 

If this is the case, then we might need to change the storage layer for quota, or else (initially) require that all ResourceQuotas in a namespace be disjoint in what they match?